### PR TITLE
feat(gatsby-plugin-checker): add more filters to filter out bad packages

### PR DESCRIPTION
**Pull request by reese**
## Description

Updating the gatsby-plugin-checker. It now filters out packages that lack a repository, have bad repo links, incorrect name formats, or lack READMEs. It also created a README in the folder to explain functionality and future features (such as automated notifications).

## Related Issues

This PR stems from issue #5211 . Note that the script still doesn't automatically create issues yet. As mentioned after PR #7062, this requires a Github API key, so the best solution would likely be to put the key in an ENV variable and then run it locally. I'll leave this for a later PR.

### CODE REVIEW

File: scripts/gatsby-plugin-checker/index.js [hunk 1]
```diff
@@ -64,6 +64,60 @@ const filterNotNotified = (packages, plugins) => {
   return packages.filter(pkg => notified.indexOf(pkg.name) < 0)
 }
 
+const removePackagesWithoutRepository = packages =>
+  packages.filter(p => hasRepository(p))
```
Comment (wardpeet) [inline scripts/gatsby-plugin-checker/index.js]: There is no need to create another function as it's only used once.

```suggestion
  packages.filter(package => !!package.links.repository)
```

File: scripts/gatsby-plugin-checker/index.js [hunk 2]
```diff
@@ -64,6 +64,60 @@ const filterNotNotified = (packages, plugins) => {
...
+
+const hasRepository = packageToCheck => {
+  if (packageToCheck.links.repository) {
+    return true
+  }
+  return false
+}
+
+const removePackagesWithBadRepoLinks = async packages => {
```
Comment (wardpeet) [inline scripts/gatsby-plugin-checker/index.js]: I'm not sure if we should remove these packages. The plugins still work and are on npm. It's not required to have a github repo.

File: scripts/gatsby-plugin-checker/index.js [hunk 3]
```diff
@@ -64,6 +64,60 @@ const filterNotNotified = (packages, plugins) => {
...
+  if (response.statusCode == 404) return false
+  return true
+}
+
+const removeBadNameFormats = packages => packages.filter(p => hasGoodName(p))
+
+const hasGoodName = pkg => {
+  const name = pkg.name
+
+  if (name.indexOf(`/`) !== -1) {
```
Comment (wardpeet) [inline scripts/gatsby-plugin-checker/index.js]: Let's rewrite this function to reduce its complexity (less nested ifs)

```js
const isScopedPackage = name.startsWith('@');
if (!isScopedPackage) {
  return startsWithAllowedPrefix(name);
}

const nameWithoutScope = name.slice(0, name.indexOf('/')
return startsWithAllowedPrefix(nameWithoutScope)
```

File: scripts/gatsby-plugin-checker/index.js [hunk 4]
```diff
@@ -64,6 +64,60 @@ const filterNotNotified = (packages, plugins) => {
...
+    const nameWithoutScope = name.split(`/`, 2)[1]
+    if (startsWithAllowedPrefix(nameWithoutScope)) return true
+  } else {
+    if (startsWithAllowedPrefix(name)) return true
+  }
+  return false
+}
+
+const startsWithAllowedPrefix = name => {
```
Comment (wardpeet) [inline scripts/gatsby-plugin-checker/index.js]:
```js
const startsWithAllowedPrefix = name => keywords.some(keyword => name.startsWith(keyword))
```

File: scripts/gatsby-plugin-checker/index.js [hunk 5]
```diff
@@ -64,6 +64,60 @@ const filterNotNotified = (packages, plugins) => {
...
+    if (name.indexOf(keyword) === 0) isGoodName = true
+  })
+  return isGoodName
+}
+
+const removePackagesWithoutReadme = packages =>
+  packages.filter(p => hasReadMe(p))
+
+const hasReadMe = pkg => {
+  if (pkg.links.homepage || pkg.readme) return true
```
Comment (wardpeet) [inline scripts/gatsby-plugin-checker/index.js]: we should try to do some extra work here to check if README.md is found in the repository if it has one like yarn does: https://github.com/yarnpkg/website/blob/master/js/src/lib/Details/index.js#L122-L131 we can just check for a 404 like you do with the git repository

**Comment by LekoArts** [comment]
Hi @reese 	👋 	
Do you have time to address Ward's comments? Would be awesome! Thanks 😊

**Comment by reese** [comment]
Yes, I'll get back to these once I get a chance! Sorry for the delay, hopefully will get to it in the next few days.

File: scripts/gatsby-plugin-checker/index.js (hunk 5 above)
Comment (reese) [reply scripts/gatsby-plugin-checker/index.js]: I just basically did a check for a 404 by checking if it was in the root folder of the project, not sure if there's a better way to do it, but I figure pretty much everyone puts the README in the root directory.

**Comment by wardpeet** [comment]
@reese I'm so sorry to let you do all this work but I think it was all for nothing. Initially I thought we were using this on www but we're not using this package anywhere. We use gatsby-source-npm-package-search.

Maybe you're using it and have a use case you want to share?

**Comment by reese** [comment]
@wardpeet This was a script that was requested in Issue #5211. The goal initially was to have this script available for manual usage. The idea is that since not all repos mention the `gatsby-plugin` keyword in their `package.json`, this script would be a way to collect those repos so that people could reach out to the maintainers and have them add the keyword if they wanted to appear on the Gatsby site. Eventually this would all be automated. That said, this issue might be stale by now, the original issue was closed so there may not be a need for it any more.

**Comment by wardpeet** [comment]
thanks @reese I was just going to circle back to you. I talked with @pieh and he said the same thing as you mentioned here in the comment so forgive me 😉. I'm testing it and let's merge this afterwards! 😄

**Pull request merged**